ledfx/effects/digitalrain2d.py

```diff
@@ -0,0 +1,204 @@
+import logging
+import random
+
+import numpy as np
+import voluptuous as vol
+
+from ledfx.color import validate_gradient
+from ledfx.effects.gradient import GradientEffect
+from ledfx.effects.twod import Twod
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Line:
+    def __init__(self, nx, color, offset, speed):
+        self.nx = nx
+        self.ny = 0
+        self.color = color
+        # per line modifyer on speed, norm 1.0
+        self.speed = speed
+        self.tail = 0.5
+        self.offset = offset
+        self.impulse_index = int(offset * 3)
+
+    def update(self, run_seconds, passed, tail, impulse):
+        # calculate how much to move
+        movement = passed / run_seconds * (1 + impulse[self.impulse_index])
+        self.tail = tail
+        # adjust for the code lines own speed
+        self.ny += movement * self.speed
+        if self.ny > (1 + self.tail):
+            return False
+        return True
+
+    def draw(self, draw, image, width, beat_osc):
+        x = int(self.nx * image.width)
+        y = int(self.ny * image.height)
+        line_width = max(1, int(image.width * (width / 100.0)))
+        tail_length = int(image.height * self.tail)
+
+        segment = (tail_length - line_width) / 10.0
+        for i in range(10):
+            y_start = y - (line_width - 1) - segment * i
+            y_end = y_start - segment
+
+            draw.line(
+                (x, y_start, x, y_end),
+                width=line_width,
+                fill=tuple((self.color * (1.0 - i * 0.09)).astype(int)),
+            )
+
+        beat_roll = beat_osc + self.offset
+        beat_roll = beat_roll - np.floor(beat_roll)
+
+        draw.line(
+            (x, y, x, y - (line_width - 1)),
+            width=line_width,
+            fill=tuple(
+                (np.array([255, 255, 255]) * (0.5 + beat_roll * 0.5)).astype(
+                    int
+                )
+            ),
+        )
+
+
+class Matrix2d(Twod, GradientEffect):
```

Comment:
I think we should call this effect class DigitalRain2D? 
matrix2d is going to absolutely confuse us in the future, it feels too raw/close to the underlying device/pixel setup.

Reply:
I forget that like EVEry time, doh